examples/nature_of_code/chp_10_neural_networks/10_04_network_animation.rs

```diff
@@ -0,0 +1,253 @@
+                self.sending = false;
+            }
+        }
+    }
+
+    // Draw line and traveling circle
+    fn display(&self, draw: &app::Draw) {
+        draw.line()
+            .start(self.a.position)
+            .end(self.b.position)
+            .color(BLACK)
+            .stroke_weight(1.0 + self.weight * 4.0);
+
+        if self.sending {
+            draw.ellipse().xy(self.sender).radius(8.0).color(BLACK);
+        }
+    }
+}
+
+struct Network {
+    // The network has a list of neurons
+    neurons: Vec<Neuron>,
+    // The Network now keeps a duplicate list of all Connection objects.
+    // This makes it easier to draw everything in this class
+    connections: Vec<Connection>,
+}
+
+impl Network {
+    fn new() -> Self {
+        Network {
+            neurons: Vec::new(),
+            connections: Vec::new(),
+        }
+    }
+
+    // We can add a Neuron
+    fn add_neuron(&mut self, n: Neuron) {
+        self.neurons.push(n);
+    }
+
+    // We can connect two Neurons
+    fn connect(&mut self, a: &mut Neuron, b: &Neuron, weight: f32) {
+        let c = Connection::new(a.clone(), b.clone(), weight);
+        // Also add the Connection here
+        self.connections.push(c.clone());
+        a.add_connection(c);
+    }
+
+    // Sending an input to the first Neuron
+    // We should do something better to track multiple inputs
+    fn feedforward(&mut self, input: f32) {
+        self.neurons[0].feedforward(input);
+    }
+
+    // Update the animation
+    fn update(&mut self) {
+        for n in &mut self.neurons {
+            n.update();
+        }
+
+        for c in &mut self.connections {
+            c.update();
+        }
+    }
+
+    // We can draw the network
+    fn display(&self, draw: &app::Draw) {
+        for n in &self.neurons {
+            n.display(&draw);
+        }
+        for c in &self.connections {
+            c.display(&draw);
+        }
+    }
+}
+
+struct Model {
+    network: Network,
+}
+
+fn model(app: &App) -> Model {
+    app.new_window()
+        .with_dimensions(640, 360)
+        .view(view)
+        .build()
+        .unwrap();
+
+    // Create an empty network
+    let mut network = Network::new();
+
+    // Create a bunch of neurons
+    let mut a = Neuron::new(-275.0, 0.0);
+    let mut b = Neuron::new(-150.0, 0.0);
+    let mut c = Neuron::new(0.0, 75.0);
+    let mut d = Neuron::new(0.0, -75.0);
+    let mut e = Neuron::new(150.0, 0.0);
+    let f = Neuron::new(275.0, 0.0);
+
+    // Connect them
+    network.connect(&mut a, &b, 1.0);
+    network.connect(&mut b, &c, random_f32());
+    network.connect(&mut b, &d, random_f32());
+    network.connect(&mut c, &e, random_f32());
+    network.connect(&mut d, &e, random_f32());
+    network.connect(&mut e, &f, 1.0);
+
+    // Add them to the network
+    network.add_neuron(a);
+    network.add_neuron(b);
+    network.add_neuron(c);
+    network.add_neuron(d);
+    network.add_neuron(e);
+    network.add_neuron(f);
```

Comment:
One possible solution would be to add the neurons *before* connecting them, and have the `add_neuron` method return the index of the neuron. Then have the `connect`method accept neuron indices (rather than references to neurons).  Does that fix the ownership issue you were running into?

Reply:
Yeah cool that makes sense but that not where my problem is. Its the fact that the "Network" has a vector of Neurons and a vector of Connections. Yet the connection struct also has needs references to two Neurons, and each Neuron needs to hold a vector of references to Connections! Phew. messy ugly stuff.....

Reply:
Ahh yep I see what you mean!

The Processing approach is disallowed in Rust by the borrow checker for a pretty good reason. E.g. it would be very easy to remove some of the connections while forgetting to also remove them from the nodes involved in those connections (or vice versa).

The reason why Processing can allow this and be memory safe is because each of the Neurons and Connections are heap-allocated objects that will get garbage collected once they are no longer referenced. This means that a neuron accessing a removed connection would still be technically memory safe. That is, the object would still be there on the heap as the neuron still owns a reference to it (even if the user has removed it). That said, this would still likely result in broken, confusing logic due to the neuron trying to update with connections that should no longer exist. At the very least there would be a memory leak.

If we wanted to match this Processing approach more closely, we could create each node and each connection as `Rc<RefCell<Node>>` and `Rc<RefCell<Connection>>` respectively, but that would probably be even more confusing to new users!

I think the more rustic approach, or a safer approach in any language, would be to make it the responsibility of the `Network` to apply the update that occurs between neurons and their connections as the `Network` is the one true owner of all the nodes and all the connections, rather than making it the responsibility of each of the neurons and connections as the web of ownership gets messy when they all refer to each other.

I wonder if we can refactor out some of the self-mutating methods on neuron and connection and move that into the network's update method? If the Connection type just stores indices into the Neurons list (rather than references to neurons) and we remove the list of connections from the Neuron type (as we already know what their connections are via the connections list), we should be able to achieve this. I haven't looked too closely just yet, but this is probably how I'd approach this!

Reply:
Great, this sounds like a good approach. I was wondering whether we should start using RefCell's but agree, it probably makes the nature of code examples a bit hairier than they need to be. I'll have a go at refactoring the code using indices in place of references to types and have the Network type handle all the logic. Cheers for the feedback.